<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Validation report</title>
     <link href="SDGSValidation/resources/css/bootstrap.min.css" rel="stylesheet">
</head>
<body>
    <div>
        <h1> Validation report </h1>
    </div>
    <div>
        <table class="table table-stripped">
  <thead>
    <tr style="text-align: left;">
      <th>File</th>
      <th>Total Variants</th>
      <th>Type</th>
      <th>Variants Fail Mismatch</th>
      <th>Variants Missing</th>
      <th>Variants Shared</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>/results/Analysis/HiSeq/2015/1504850/S0902255-07/Ataxia/1504850-S0902255-07_ConsensusVariants_Ataxia_LessLQs.txt</td>
      <td>91</td>
      <td>baseline</td>
      <td>-</td>
      <td>29</td>
      <td>72</td>
    </tr>
    <tr>
      <td>/home/bioinfo/mparker/work/20160613/pipeline/1504850-S0902255-07_Ataxia_variants_with_downsampling.txt</td>
      <td>101</td>
      <td>test</td>
      <td>3</td>
      <td>19</td>
      <td>72</td>
    </tr>
  </tbody>
</table>
    </div>
    <div>
        <h2>Missing Variants</h2>
    </div>
    <div>
        <table border="1" class="dataframe">
  <thead>
    <tr style="text-align: left;">
      <th>alt</th>
      <th>chrom</th>
      <th>filter</th>
      <th>genotype</th>
      <th>id</th>
      <th>info</th>
      <th>key</th>
      <th>pos</th>
      <th>qual</th>
      <th>ref</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>-</td>
      <td>chr5</td>
      <td>PASS</td>
      <td>0/1:99:193,113:306:2144,0,4835</td>
      <td>rs78316512</td>
      <td>exons=3,2;genes=SLC1A3,SLC1A3;consequence=splicing;cDNA_changes=c.182-22T&gt;-,c.182-22T&gt;-;transcripts=NM_004172,NM_001166695;protein_changes=NA,NA</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>36629530</td>
      <td>1261.73</td>
      <td>T</td>
    </tr>
    <tr>
      <td>A</td>
      <td>chr9</td>
      <td>PASS</td>
      <td>1/1:99:25,321:346:10601,713,0</td>
      <td>rs66524976</td>
      <td>exons=;genes=;consequence=splicing;cDNA_changes=;transcripts=;protein_changes=</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>135229017</td>
      <td>2.14748e+09</td>
      <td>-</td>
    </tr>
    <tr>
      <td>CTG</td>
      <td>chr19</td>
      <td>PASS</td>
      <td>0/1:99:123,95:218:2512,0,4886</td>
      <td>.</td>
      <td>exons=47;genes=CACNA1A;consequence=nonframeshift insertion;cDNA_changes=c.6975_6976insCAG;transcripts=NM_001127222;protein_changes=p.A2326delinsQA</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>13318672</td>
      <td>2474.73</td>
      <td>-</td>
    </tr>
    <tr>
      <td>-</td>
      <td>chr2</td>
      <td>PASS</td>
      <td>1/2:99:7,192,129:328:11892,4054,3725,5328,0,4664</td>
      <td>rs201455607</td>
      <td>exons=;genes=;consequence=splicing;cDNA_changes=;transcripts=;protein_changes=</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>166895914</td>
      <td>4.29497e+09</td>
      <td>TTT</td>
    </tr>
    <tr>
      <td>-</td>
      <td>chr9</td>
      <td>PASS</td>
      <td>0/1:99:314,60:374:351,0,9328</td>
      <td>rs34769225</td>
      <td>exons=23;genes=SETX;consequence=splicing;cDNA_changes=c.6843-5T&gt;-;transcripts=NM_015046;protein_changes=NA</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>135152544</td>
      <td>313.73</td>
      <td>A</td>
    </tr>
    <tr>
      <td>-</td>
      <td>chr9</td>
      <td>PASS</td>
      <td>1/2:99:50,132,120:302:5656,1712,2437,1970,0,3694</td>
      <td>.</td>
      <td>exons=9,10;genes=APTX,APTX;consequence=splicing;cDNA_changes=c.917-25T&gt;-,c.875-25T&gt;-;transcripts=NM_175069,NM_001195251;protein_changes=NA,NA</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>32973711</td>
      <td>3570.19</td>
      <td>A</td>
    </tr>
    <tr>
      <td>A</td>
      <td>chr5</td>
      <td>PASS</td>
      <td>0/1:99:179,157:336:4747,0,4956</td>
      <td>rs3216728</td>
      <td>exons=2,1;genes=SLC1A3,SLC1A3;consequence=splicing;cDNA_changes=c.181+21-&gt;A,c.181+21-&gt;A;transcripts=NM_004172,NM_001166695;protein_changes=NA,NA</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>36608727</td>
      <td>4709.73</td>
      <td>-</td>
    </tr>
    <tr>
      <td>GGTGAGAAAT</td>
      <td>chr1</td>
      <td>PASS</td>
      <td>0/1:99:175,151:326:5880,0,17641</td>
      <td>rs11282849</td>
      <td>exons=;genes=SLC2A1;consequence=intronic;cDNA_changes=;transcripts=;protein_changes=</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>43393245</td>
      <td>5842.73</td>
      <td>-</td>
    </tr>
    <tr>
      <td>A</td>
      <td>chr9</td>
      <td>PASS</td>
      <td>1/1:99:12,387:399:13311,992,0</td>
      <td>rs34600530</td>
      <td>exons=7,8,9,8,8,9,7,8;genes=APTX,APTX,APTX,APTX,APTX,APTX,APTX,APTX;consequence=splicing;cDNA_changes=c.609-11-&gt;T,c.597-11-&gt;T,c.771-11-&gt;T,c.813-11-&gt;T,c.813-11-&gt;T,c.771-11-&gt;T,c.651-11-&gt;T,c.771-11-&gt;T;transcripts=NM_001195254,NM_001195252,NM_175073,NM_175069,NM_001195248,NM_001195251,NM_001195250,NM_001195249;protein_changes=NA,NA,NA,NA,NA,NA,NA,NA</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>32974570</td>
      <td>2.14748e+09</td>
      <td>-</td>
    </tr>
    <tr>
      <td>-</td>
      <td>chr11</td>
      <td>PASS</td>
      <td>0/1:99:262,56:318:356,0,7573</td>
      <td>.</td>
      <td>exons=6;genes=ATM;consequence=splicing;cDNA_changes=c.497-18T&gt;-;transcripts=NM_000051;protein_changes=NA</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>108114662</td>
      <td>318.73</td>
      <td>T</td>
    </tr>
    <tr>
      <td>-</td>
      <td>chr11</td>
      <td>PASS</td>
      <td>0/1:99:216,44:260:193,0,6140</td>
      <td>.</td>
      <td>exons=10;genes=ATM;consequence=splicing;cDNA_changes=c.1236-17T&gt;-;transcripts=NM_000051;protein_changes=NA</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>108121411</td>
      <td>155.73</td>
      <td>T</td>
    </tr>
    <tr>
      <td>-</td>
      <td>chr11</td>
      <td>PASS</td>
      <td>0/1:99:287,58:345:842,0,7715</td>
      <td>rs373881770</td>
      <td>exons=20;genes=ATM;consequence=splicing;cDNA_changes=c.2922-22T&gt;-;transcripts=NM_000051;protein_changes=NA</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>108141956</td>
      <td>804.73</td>
      <td>T</td>
    </tr>
    <tr>
      <td>-</td>
      <td>chr8</td>
      <td>PASS</td>
      <td>1/1:99:0,393:393:14299,1189,0</td>
      <td>rs5891887</td>
      <td>exons=4;genes=TTPA;consequence=splicing;cDNA_changes=c.359-3T&gt;-;transcripts=NM_000370;protein_changes=NA</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>63978659</td>
      <td>14261.7</td>
      <td>A</td>
    </tr>
    <tr>
      <td>-</td>
      <td>chr9</td>
      <td>PASS</td>
      <td>./.:.:.:.:.</td>
      <td>rs376402105</td>
      <td>exons=5,6,7,6,6,7,5,6;genes=APTX,APTX,APTX,APTX,APTX,APTX,APTX,APTX;consequence=splicing;cDNA_changes=c.322-3T&gt;-,c.310-3T&gt;-,c.484-3T&gt;-,c.526-3T&gt;-,c.526-3T&gt;-,c.484-3T&gt;-,c.364-3T&gt;-,c.484-3T&gt;-;transcripts=NM_001195254,NM_001195252,NM_175073,NM_175069,NM_001195248,NM_001195251,NM_001195250,NM_001195249;protein_changes=NA,NA,NA,NA,NA,NA,NA,NA</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>32986031</td>
      <td>161.73</td>
      <td>A</td>
    </tr>
    <tr>
      <td>-</td>
      <td>chr1</td>
      <td>PASS</td>
      <td>1/1:99:3,337:340:14423,1014,0</td>
      <td>rs67915591</td>
      <td>exons=;genes=;consequence=splicing;cDNA_changes=;transcripts=;protein_changes=</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>160090682</td>
      <td>14385.7</td>
      <td>CCTT</td>
    </tr>
    <tr>
      <td>-</td>
      <td>chr16</td>
      <td>PASS</td>
      <td>0/1:99:340</td>
      <td>rs5818722</td>
      <td>exons=2,2;genes=SPG7,SPG7;consequence=splicing;cDNA_changes=c.184-18T&gt;-,c.184-18T&gt;-;transcripts=NM_199367,NM_003119;protein_changes=NA,NA</td>
      <td>GT:GQ:DP</td>
      <td>89576880</td>
      <td>349.73</td>
      <td>T</td>
    </tr>
    <tr>
      <td>AAA</td>
      <td>chr10</td>
      <td>PASS</td>
      <td>1/1:26:35</td>
      <td>rs61583906</td>
      <td>exons=;genes=MTPAP;consequence=UTR5;cDNA_changes=;transcripts=;protein_changes=</td>
      <td>GT:GQ:DP</td>
      <td>30638208</td>
      <td>97.83</td>
      <td>-</td>
    </tr>
    <tr>
      <td>C</td>
      <td>chr9</td>
      <td>PASS</td>
      <td>0/1:99:181,171:352:6595,0,7138</td>
      <td>rs3833700</td>
      <td>exons=12;genes=GBA2;consequence=splicing;cDNA_changes=c.1688-9-&gt;G;transcripts=NM_020944;protein_changes=NA</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>35739115</td>
      <td>6557.73</td>
      <td>-</td>
    </tr>
    <tr>
      <td>-</td>
      <td>chr11</td>
      <td>PASS</td>
      <td>0/1:99:298,39:337:288,0,8537</td>
      <td>rs58978479</td>
      <td>exons=43;genes=ATM;consequence=splicing;cDNA_changes=c.6347+19T&gt;-;transcripts=NM_000051;protein_changes=NA</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>108188267</td>
      <td>250.73</td>
      <td>T</td>
    </tr>
  </tbody>
</table>
    </div>
    <div>
        <h2>Mismatch Variants</h2>
    </div>
    <div>
        <table border="1" class="dataframe">
  <thead>
    <tr style="text-align: left;">
      <th>scope</th>
      <th>chrom</th>
      <th>pos</th>
      <th>ref</th>
      <th>alt</th>
      <th>key</th>
      <th>genotype</th>
      <th>id</th>
      <th>qual</th>
      <th>info</th>
      <th>bam</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>BASELINE</td>
      <td>chr12</td>
      <td>5022041</td>
      <td>G</td>
      <td>A</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>0/1:99:177,172:349:5322,0,5402</td>
      <td>rs4766310</td>
      <td>5293.77</td>
      <td>exons=<br>genes=KCNA1<br>consequence=UTR3<br>cDNA_changes=<br>transcripts=<br>protein_changes=</td>
      <td></td>
    </tr>
    <tr>
      <td>TESTING</td>
      <td>chr12</td>
      <td>5022041</td>
      <td>G</td>
      <td>A</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>0/1:99:116,109:225:3356,0,3174</td>
      <td>rs4766310</td>
      <td>3327.77</td>
      <td>MLEAC=1<br>AC=1<br>BaseQRankSum=0.974<br>MLEAF=0.500<br>exons=UTR3<br>ReadPosRankSum=0.005<br>AF=0.500<br>genes=KCNA1<br>AN=2<br>MQ0=0<br>FS=0.000<br>MQ=60.00<br>DP=228<br>cDNA_changes=NA<br>MQRankSum=-0.056<br>transcripts=NM_000217<br>dataset=full<br>protein_changes=NA<br>ClippingRankSum=-0.368<br>QD=14.60</td>
      <td><a href="http://localhost:60151/load?file=http://10.182.131.21/results/Analysis/HiSeq/2016/1600553/S0901823-02/1600553-S0901823-02_Aligned_Sorted_PCRDuped_IndelsRealigned.bam,http://10.182.131.21/results/Analysis/HiSeq/2016/1600553/S1414882-04/1600553-S1414882-04_Aligned_Sorted_PCRDuped_IndelsRealigned.bam&merge=false">BamView</a></td>
    </tr>
    <tr>
      <td>BASELINE</td>
      <td>chr9</td>
      <td>32986040</td>
      <td>A</td>
      <td>C</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>./.:.:.:.:.</td>
      <td>rs1981011</td>
      <td>359.77</td>
      <td>exons=5,6,7,6,6,7,5,6<br>genes=APTX,APTX,APTX,APTX,APTX,APTX,APTX,APTX<br>consequence=splicing<br>cDNA_changes=c.322-12T>G,c.310-12T>G,c.484-12T>G,c.526-12T>G,c.526-12T>G,c.484-12T>G,c.364-12T>G,c.484-12T>G<br>transcripts=NM_001195254,NM_001195252,NM_175073,NM_175069,NM_001195248,NM_001195251,NM_001195250,NM_001195249<br>protein_changes=NA,NA,NA,NA,NA,NA,NA,NA</td>
      <td></td>
    </tr>
    <tr>
      <td>TESTING</td>
      <td>chr9</td>
      <td>32986040</td>
      <td>A</td>
      <td>C</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>0/1:99:2,13:15:466,0,322</td>
      <td>rs1981011</td>
      <td>437.77</td>
      <td>MLEAC=1<br>AC=1<br>BaseQRankSum=0.764<br>MLEAF=0.500<br>exons=5,6,7,6,6,7,5,6<br>ReadPosRankSum=0.594<br>AF=0.500<br>genes=APTX<br>AN=2<br>MQ0=0<br>FS=0.000<br>MQ=60.00<br>DP=17<br>cDNA_changes=c.322-12T>G,c.310-12T>G,c.484-12T>G,c.526-12T>G,c.526-12T>G,c.484-12T>G,c.364-12T>G,c.484-12T>G<br>MQRankSum=0.764<br>transcripts=NM_001195254,NM_001195252,NM_175073,NM_175069,NM_001195248,NM_001195251,NM_001195250,NM_001195249<br>dataset=quarter<br>protein_changes=NA<br>ClippingRankSum=0.255<br>QD=25.75</td>
      <td><a href="http://localhost:60151/load?file=http://10.182.131.21/results/Analysis/HiSeq/2016/1600553/S0901823-02/1600553-S0901823-02_Aligned_Sorted_PCRDuped_IndelsRealigned.bam,http://10.182.131.21/results/Analysis/HiSeq/2016/1600553/S1414882-04/1600553-S1414882-04_Aligned_Sorted_PCRDuped_IndelsRealigned.bam&merge=false">BamView</a></td>
    </tr>
    <tr>
      <td>BASELINE</td>
      <td>chr2</td>
      <td>166897989</td>
      <td>G</td>
      <td>A</td>
      <td>GT:GQ:DP</td>
      <td>1/2:99:254</td>
      <td>rs199507113</td>
      <td>1501.19</td>
      <td>exons=14,14,16,14<br>genes=SCN1A,SCN1A,SCN1A,SCN1A<br>consequence=splicing<br>cDNA_changes=c.2144-10C>T,c.2093-10C>T,c.2177-10C>T,c.2177-10C>T<br>transcripts=NM_006920,NM_001165964,NM_001202435,NM_001165963<br>protein_changes=NA,NA,NA,NA</td>
      <td></td>
    </tr>
    <tr>
      <td>TESTING</td>
      <td>chr2</td>
      <td>166897989</td>
      <td>G</td>
      <td>GA</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>0/1:99:37,135:172:3311,0,680</td>
      <td>rs11394960</td>
      <td>1315.73</td>
      <td>MLEAC=1<br>AC=1<br>BaseQRankSum=-0.726<br>MLEAF=0.500<br>exons=14,14,16,14<br>ReadPosRankSum=0.300<br>AF=0.500<br>genes=SCN1A<br>AN=2<br>MQ0=0<br>FS=3.744<br>MQ=60.00<br>DP=224<br>cDNA_changes=c.2144-10->T,c.2093-10->T,c.2177-10->T,c.2177-10->T<br>MQRankSum=0.319<br>transcripts=NM_006920,NM_001165964,NM_001202435,NM_001165963<br>dataset=full<br>protein_changes=NA<br>ClippingRankSum=-0.274<br>QD=5.87</td>
      <td><a href="http://localhost:60151/load?file=http://10.182.131.21/results/Analysis/HiSeq/2016/1600553/S0901823-02/1600553-S0901823-02_Aligned_Sorted_PCRDuped_IndelsRealigned.bam,http://10.182.131.21/results/Analysis/HiSeq/2016/1600553/S1414882-04/1600553-S1414882-04_Aligned_Sorted_PCRDuped_IndelsRealigned.bam&merge=false">BamView</a></td>
    </tr>
  </tbody>
</table>
    </div>
    <div>
        <h2>Additional Variants</h2>
    </div>
    <div>
        <table border="1" class="dataframe">
  <thead>
    <tr style="text-align: left;">
      <th>alt</th>
      <th>chrom</th>
      <th>filter</th>
      <th>genotype</th>
      <th>id</th>
      <th>info</th>
      <th>key</th>
      <th>pos</th>
      <th>qual</th>
      <th>ref</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>GA</td>
      <td>chr9</td>
      <td>PASS</td>
      <td>1/1:99:10,295:305:10016,752,0</td>
      <td>rs34600530</td>
      <td>MLEAC=2;AC=2;BaseQRankSum=-1.189;MLEAF=1.00;exons=7,8,9,8,8,9,7,8;ReadPosRankSum=-1.328;AF=1.00;genes=APTX;AN=2;MQ0=0;FS=1.504;MQ=60.00;DP=349;cDNA_changes=c.609-11-&gt;T,c.597-11-&gt;T,c.771-11-&gt;T,c.813-11-&gt;T,c.813-11-&gt;T,c.771-11-&gt;T,c.651-11-&gt;T,c.771-11-&gt;T;MQRankSum=0.169;transcripts=NM_001195254,NM_001195252,NM_175073,NM_175069,NM_001195248,NM_001195251,NM_001195250,NM_001195249;dataset=full;protein_changes=NA;ClippingRankSum=0.569;QD=25.11</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>32974570</td>
      <td>2.14748E9</td>
      <td>G</td>
    </tr>
    <tr>
      <td>CA</td>
      <td>chr5</td>
      <td>PASS</td>
      <td>0/1:99:142,125:267:3790,0,3915</td>
      <td>rs3216728</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=-0.485;MLEAF=0.500;exons=2,1,UTR3;ReadPosRankSum=-0.358;AF=0.500;genes=SLC1A3;AN=2;MQ0=0;FS=0.633;MQ=60.07;DP=300;cDNA_changes=c.181+21-&gt;A,c.181+21-&gt;A,NA;MQRankSum=0.601;transcripts=NM_004172,NM_001166695,NM_001166696;dataset=full;protein_changes=NA;ClippingRankSum=0.426;QD=12.51</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>36608727</td>
      <td>3752.73</td>
      <td>C</td>
    </tr>
    <tr>
      <td>T</td>
      <td>chr16</td>
      <td>PASS</td>
      <td>0/1:43:218,76:294:43,0,7284</td>
      <td>.</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=-12.245;MLEAF=0.500;exons=11;ReadPosRankSum=-2.884;AF=0.500;genes=SPG7;AN=2;MQ0=0;FS=437.483;MQ=60.00;DP=295;cDNA_changes=c.1468G&gt;T;MQRankSum=-0.601;transcripts=NM_003119;dataset=full;protein_changes=p.E490X;ClippingRankSum=9.633;QD=0.05</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>89613084</td>
      <td>14.91</td>
      <td>G</td>
    </tr>
    <tr>
      <td>A</td>
      <td>chr11</td>
      <td>PASS</td>
      <td>0/1:99:239,42:281:170,0,6834</td>
      <td>.</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=0.351;MLEAF=0.500;exons=6;ReadPosRankSum=-1.325;AF=0.500;genes=ATM;AN=2;MQ0=0;FS=0.000;MQ=60.07;DP=325;cDNA_changes=c.497-18T&gt;-;MQRankSum=-0.519;transcripts=NM_000051;dataset=full;protein_changes=NA;ClippingRankSum=1.319;QD=0.41</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>108114661</td>
      <td>132.73</td>
      <td>AT</td>
    </tr>
    <tr>
      <td>C</td>
      <td>chr1</td>
      <td>PASS</td>
      <td>1/1:99:0,257:257:11070,791,0</td>
      <td>rs67915591</td>
      <td>MLEAC=2;AC=2;FS=0.000;MLEAF=1.00;exons=in_25_bp_from_exon;AF=1.00;genes=ATP1A2;AN=2;MQ0=0;MQ=61.30;QD=10.45;cDNA_changes=NA;transcripts=NM_000702;dataset=full;protein_changes=NA;DP=264</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>160090681</td>
      <td>11032.7</td>
      <td>CCCTT</td>
    </tr>
    <tr>
      <td>CCTG</td>
      <td>chr19</td>
      <td>PASS</td>
      <td>0/1:99:81,62:143:2170,0,3219</td>
      <td>.</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=2.938;MLEAF=0.500;exons=47,unknown_exonic,UTR3,UTR3,UTR3;ReadPosRankSum=-1.089;AF=0.500;genes=CACNA1A;AN=2;MQ0=0;FS=6.332;MQ=60.21;DP=236;cDNA_changes=c.6975_6976insCAG,NA,NA,NA,NA;MQRankSum=-0.426;transcripts=NM_001127222,NM_023035,NM_000068,NM_001127221,NM_001174080;dataset=full;protein_changes=p.A2326delinsQA,NA,NA,NA,NA;ClippingRankSum=-0.135;QD=3.01</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>13318672</td>
      <td>2132.73</td>
      <td>C</td>
    </tr>
    <tr>
      <td>A</td>
      <td>chr3</td>
      <td>PASS</td>
      <td>0/1:31:77,7:84:31,0,8169</td>
      <td>.</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=-4.370;MLEAF=0.500;exons=37,37,38;ReadPosRankSum=2.460;AF=0.500;genes=ITPR1;AN=2;MQ0=0;FS=87.529;MQ=60.00;DP=84;cDNA_changes=c.4919+5G&gt;A,c.4946+5G&gt;A,c.4964+5G&gt;A;MQRankSum=-0.518;transcripts=NM_002222,NM_001168272,NM_001099952;dataset=quarter;protein_changes=NA;ClippingRankSum=-3.188;QD=0.05</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>4752162</td>
      <td>4.61</td>
      <td>G</td>
    </tr>
    <tr>
      <td>A</td>
      <td>chr8</td>
      <td>PASS</td>
      <td>1/1:99:15,335:350:12698,1008,0</td>
      <td>rs7006757</td>
      <td>MLEAC=2;AC=2;BaseQRankSum=0.819;MLEAF=1.00;exons=8,8;ReadPosRankSum=0.897;AF=1.00;genes=DDHD2;AN=2;MQ0=0;FS=0.000;MQ=59.95;DP=355;cDNA_changes=c.849-21T&gt;A,c.849-21T&gt;A;MQRankSum=1.951;transcripts=NM_015214,NM_001164232;dataset=full;protein_changes=NA;ClippingRankSum=-0.190;QD=29.33</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>38103239</td>
      <td>12669.8</td>
      <td>T</td>
    </tr>
    <tr>
      <td>T</td>
      <td>chr8</td>
      <td>PASS</td>
      <td>1/1:99:0,327:327:11895,990,0</td>
      <td>rs5891887</td>
      <td>MLEAC=2;AC=2;FS=0.000;MLEAF=1.00;exons=4;AF=1.00;genes=TTPA;AN=2;MQ0=0;MQ=60.15;QD=33.31;cDNA_changes=c.359-3T&gt;-;transcripts=NM_000370;dataset=full;protein_changes=NA;DP=356</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>63978658</td>
      <td>11857.7</td>
      <td>TA</td>
    </tr>
    <tr>
      <td>C</td>
      <td>chr3</td>
      <td>PASS</td>
      <td>0/1:75:250,80:330:75,0,7868</td>
      <td>.</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=-12.942;MLEAF=0.500;exons=20,21,20;ReadPosRankSum=5.694;AF=0.500;genes=ITPR1;AN=2;MQ0=0;FS=446.380;MQ=60.00;DP=335;cDNA_changes=c.2353A&gt;C,c.2398A&gt;C,c.2353A&gt;C;MQRankSum=-0.421;transcripts=NM_001168272,NM_001099952,NM_002222;dataset=full;protein_changes=p.T785P,p.T800P,p.T785P;ClippingRankSum=-4.402;QD=0.14</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>4715013</td>
      <td>46.77</td>
      <td>A</td>
    </tr>
    <tr>
      <td>TA</td>
      <td>chr9</td>
      <td>PASS</td>
      <td>1/1:99:15,231:246:7633,435,0</td>
      <td>rs66524976</td>
      <td>MLEAC=2;AC=2;BaseQRankSum=-0.636;MLEAF=1.00;exons=in_25_bp_from_exon;ReadPosRankSum=0.865;AF=1.00;genes=SETX;AN=2;MQ0=0;FS=0.000;MQ=60.12;DP=274;cDNA_changes=NA;MQRankSum=0.343;transcripts=NM_015046;dataset=full;protein_changes=NA;ClippingRankSum=-1.198;QD=23.13</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>135229017</td>
      <td>2.14748E9</td>
      <td>T</td>
    </tr>
    <tr>
      <td>T</td>
      <td>chr8</td>
      <td>PASS</td>
      <td>1/1:99:13,419:432:15761,1258,0</td>
      <td>rs2306899</td>
      <td>MLEAC=2;AC=2;BaseQRankSum=1.509;MLEAF=1.00;exons=5,5,5;ReadPosRankSum=0.778;AF=1.00;genes=DDHD2;AN=2;MQ0=0;FS=1.084;MQ=60.00;DP=433;cDNA_changes=c.557C&gt;T,c.557C&gt;T,c.557C&gt;T;MQRankSum=1.888;transcripts=NM_015214,NM_001164234,NM_001164232;dataset=full;protein_changes=p.T186M,p.T186M,p.T186M;ClippingRankSum=0.341;QD=26.56</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>38095662</td>
      <td>15732.8</td>
      <td>C</td>
    </tr>
    <tr>
      <td>C</td>
      <td>chr11</td>
      <td>PASS</td>
      <td>0/1:99:182,40:222:532,0,4714</td>
      <td>.</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=0.080;MLEAF=0.500;exons=10;ReadPosRankSum=-0.530;AF=0.500;genes=ATM;AN=2;MQ0=0;FS=0.971;MQ=60.08;DP=267;cDNA_changes=c.1236-17T&gt;-;MQRankSum=-0.484;transcripts=NM_000051;dataset=full;protein_changes=NA;ClippingRankSum=0.491;QD=1.85</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>108121410</td>
      <td>494.73</td>
      <td>CT</td>
    </tr>
    <tr>
      <td>C</td>
      <td>chr11</td>
      <td>PASS</td>
      <td>0/1:99:281,64:345:594,0,7514</td>
      <td>rs373881770</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=-0.036;MLEAF=0.500;exons=20;ReadPosRankSum=-0.170;AF=0.500;genes=ATM;AN=2;MQ0=0;FS=1.913;MQ=60.03;DP=430;cDNA_changes=c.2922-22T&gt;-;MQRankSum=-0.454;transcripts=NM_000051;dataset=full;protein_changes=NA;ClippingRankSum=-0.191;QD=0.69</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>108141955</td>
      <td>298.73</td>
      <td>CT</td>
    </tr>
    <tr>
      <td>C</td>
      <td>chr9</td>
      <td>PASS</td>
      <td>1/2:99:41,105,93:239:4501,1168,1794,1881,0,3274</td>
      <td>.</td>
      <td>BaseQRankSum=0.329;exons=9,10,intronic,intronic,intronic,intronic,intronic,intronic;dataset=full;FS=0.807;orig_alt=CA,C;DP=313;MLEAC=1,1;MLEAF=0.500,0.500;orig_ref=CAA;cDNA_changes=c.917-24T&gt;-,c.875-24T&gt;-,NA,NA,NA,NA,NA,NA;ClippingRankSum=-0.637;AC=1,1;MQRankSum=0.910;AF=0.500,0.500;genes=APTX;AN=2;MQ0=0;protein_changes=NA;ReadPosRankSum=0.375;MQ=60.17;QD=6.26;transcripts=NM_175069,NM_001195251,NM_001195248,NM_001195249,NM_001195250,NM_001195252,NM_001195254,NM_175073</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>32973709</td>
      <td>2937.19</td>
      <td>CA</td>
    </tr>
    <tr>
      <td>C</td>
      <td>chr2</td>
      <td>PASS</td>
      <td>1/2:99:4,161,102:267:9630,3189,2741,4452,0,3886</td>
      <td>rs201455607</td>
      <td>BaseQRankSum=1.497;exons=in_25_bp_from_exon;dataset=full;FS=5.172;orig_alt=C,CT;DP=340;MLEAC=1,1;MLEAF=0.500,0.500;orig_ref=CTTT;cDNA_changes=NA;ClippingRankSum=-0.323;AC=1,1;MQRankSum=-0.866;AF=0.500,0.500;genes=SCN1A;AN=2;MQ0=0;protein_changes=NA;ReadPosRankSum=0.787;MQ=60.13;QD=26.88;transcripts=NM_001165963,NM_001165964,NM_001202435,NM_006920</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>166895913</td>
      <td>2.14749E9</td>
      <td>CTT</td>
    </tr>
    <tr>
      <td>C</td>
      <td>chr8</td>
      <td>PASS</td>
      <td>1/1:99:23,292:315:9865,704,0</td>
      <td>rs2898674</td>
      <td>MLEAC=2;AC=2;BaseQRankSum=1.608;MLEAF=1.00;exons=14,14;ReadPosRankSum=-0.244;AF=1.00;genes=DDHD2;AN=2;MQ0=0;FS=3.580;MQ=60.06;DP=513;cDNA_changes=c.1618-25T&gt;C,c.1618-25T&gt;C;MQRankSum=0.992;transcripts=NM_015214,NM_001164232;dataset=full;protein_changes=NA;ClippingRankSum=-4.197;QD=28.43</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>38110215</td>
      <td>2.14748E9</td>
      <td>T</td>
    </tr>
    <tr>
      <td>GGGTGAGAAAT</td>
      <td>chr1</td>
      <td>PASS</td>
      <td>0/1:99:104,127:231:4571,0,11463</td>
      <td>rs11282849</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=-0.462;MLEAF=0.500;exons=intronic;ReadPosRankSum=-3.203;AF=0.500;genes=SLC2A1;AN=2;MQ0=0;FS=13.904;MQ=60.00;DP=240;cDNA_changes=NA;MQRankSum=0.901;transcripts=NM_006516;dataset=full;protein_changes=NA;ClippingRankSum=0.789;QD=1.89</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>43393245</td>
      <td>4533.73</td>
      <td>G</td>
    </tr>
    <tr>
      <td>C</td>
      <td>chr14</td>
      <td>PASS</td>
      <td>1/1:99:0,393:393:14881,1184,0</td>
      <td>rs3742883</td>
      <td>MLEAC=2;AC=2;FS=0.000;MLEAF=1.00;exons=31;AF=1.00;genes=ZFYVE26;AN=2;MQ0=0;MQ=60.00;QD=25.45;cDNA_changes=c.5672A&gt;G;transcripts=NM_015346;dataset=full;protein_changes=p.N1891S;DP=394</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>68234539</td>
      <td>14852.8</td>
      <td>T</td>
    </tr>
    <tr>
      <td>C</td>
      <td>chr2</td>
      <td>PASS</td>
      <td>1/2:99:4,161,102:267:9630,3189,2741,4452,0,3886</td>
      <td>rs201455607</td>
      <td>dataset=full;MLEAC=1,1;AC=1,1;BaseQRankSum=1.497;MLEAF=0.500,0.500;exons=in_25_bp_from_exon;ReadPosRankSum=0.787;AF=0.500,0.500;genes=SCN1A;AN=2;MQ0=0;FS=5.172;MQ=60.13;DP=340;cDNA_changes=NA;MQRankSum=-0.866;transcripts=NM_001165963,NM_001165964,NM_001202435,NM_006920;orig_alt=C,CT;protein_changes=NA;ClippingRankSum=-0.323;QD=26.88</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>166895913</td>
      <td>2.14749E9</td>
      <td>CTTT</td>
    </tr>
    <tr>
      <td>C</td>
      <td>chr9</td>
      <td>PASS</td>
      <td>1/2:99:41,105,93:239:4501,1168,1794,1881,0,3274</td>
      <td>.</td>
      <td>dataset=full;MLEAC=1,1;AC=1,1;BaseQRankSum=0.329;MLEAF=0.500,0.500;exons=in_25_bp_from_exon,in_25_bp_from_exon,intronic,intronic,intronic,intronic,intronic,intronic;ReadPosRankSum=0.375;AF=0.500,0.500;genes=APTX;AN=2;MQ0=0;FS=0.807;MQ=60.17;DP=313;cDNA_changes=NA;MQRankSum=0.910;transcripts=NM_001195251,NM_175069,NM_001195248,NM_001195249,NM_001195250,NM_001195252,NM_001195254,NM_175073;orig_alt=CA,C;protein_changes=NA;ClippingRankSum=-0.637;QD=6.26</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>32973709</td>
      <td>2937.19</td>
      <td>CAA</td>
    </tr>
    <tr>
      <td>G</td>
      <td>chr16</td>
      <td>PASS</td>
      <td>0/1:99:238,47:285:545,0,6102</td>
      <td>rs5818722</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=0.719;MLEAF=0.500;exons=2,2;ReadPosRankSum=1.010;AF=0.500;genes=SPG7;AN=2;MQ0=0;FS=3.111;MQ=60.00;DP=336;cDNA_changes=c.184-18T&gt;-,c.184-18T&gt;-;MQRankSum=0.429;transcripts=NM_199367,NM_003119;dataset=full;protein_changes=NA;ClippingRankSum=-1.074;QD=1.51</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>89576879</td>
      <td>507.73</td>
      <td>GT</td>
    </tr>
    <tr>
      <td>C</td>
      <td>chr5</td>
      <td>PASS</td>
      <td>0/1:99:145,85:230:1664,0,3504</td>
      <td>rs78316512</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=-0.657;MLEAF=0.500;exons=3,2;ReadPosRankSum=-0.031;AF=0.500;genes=SLC1A3;AN=2;MQ0=0;FS=12.956;MQ=60.19;DP=284;cDNA_changes=c.182-22T&gt;-,c.182-22T&gt;-;MQRankSum=1.240;transcripts=NM_004172,NM_001166695;dataset=full;protein_changes=NA;ClippingRankSum=-0.415;QD=3.48</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>36629529</td>
      <td>987.73</td>
      <td>CT</td>
    </tr>
    <tr>
      <td>G</td>
      <td>chr14</td>
      <td>PASS</td>
      <td>1/1:99:10,389:399:14413,1167,0</td>
      <td>rs12891164</td>
      <td>MLEAC=2;AC=2;BaseQRankSum=0.068;MLEAF=1.00;exons=11;ReadPosRankSum=0.368;AF=1.00;genes=ZFYVE26;AN=2;MQ0=0;FS=2.779;MQ=60.00;DP=399;cDNA_changes=c.2112T&gt;C;MQRankSum=-0.004;transcripts=NM_015346;dataset=full;protein_changes=p.P704P;ClippingRankSum=-0.237;QD=29.62</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>68264867</td>
      <td>14384.8</td>
      <td>A</td>
    </tr>
    <tr>
      <td>T</td>
      <td>chr9</td>
      <td>PASS</td>
      <td>0/1:95:41,12:53:95,0,1044</td>
      <td>rs376402105</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=0.436;MLEAF=0.500;exons=5,6,7,6,6,7,5,6;ReadPosRankSum=1.160;AF=0.500;genes=APTX;AN=2;MQ0=0;FS=6.629;MQ=60.00;DP=135;cDNA_changes=c.322-3T&gt;-,c.310-3T&gt;-,c.484-3T&gt;-,c.526-3T&gt;-,c.526-3T&gt;-,c.484-3T&gt;-,c.364-3T&gt;-,c.484-3T&gt;-;MQRankSum=-0.414;transcripts=NM_001195254,NM_001195252,NM_175073,NM_175069,NM_001195248,NM_001195251,NM_001195250,NM_001195249;dataset=full;protein_changes=NA;ClippingRankSum=-3.645;QD=0.43</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>32986030</td>
      <td>57.73</td>
      <td>TA</td>
    </tr>
    <tr>
      <td>GC</td>
      <td>chr9</td>
      <td>PASS</td>
      <td>0/1:99:139,131:270:5089,0,4938</td>
      <td>rs3833700</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=0.984;MLEAF=0.500;exons=12;ReadPosRankSum=2.079;AF=0.500;genes=GBA2;AN=2;MQ0=0;FS=2.098;MQ=60.12;DP=274;cDNA_changes=c.1688-9-&gt;G;MQRankSum=0.463;transcripts=NM_020944;dataset=full;protein_changes=NA;ClippingRankSum=-2.414;QD=18.44</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>35739115</td>
      <td>5051.73</td>
      <td>G</td>
    </tr>
    <tr>
      <td>C</td>
      <td>chr4</td>
      <td>PASS</td>
      <td>0/1:99:126,136:262:3884,0,4116</td>
      <td>rs6856745</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=-1.739;MLEAF=0.500;exons=3;ReadPosRankSum=-0.078;AF=0.500;genes=CYP2U1;AN=2;MQ0=0;FS=6.561;MQ=60.00;DP=267;cDNA_changes=c.1127-17T&gt;C;MQRankSum=1.057;transcripts=NM_183075;dataset=full;protein_changes=NA;ClippingRankSum=-0.071;QD=14.44</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>108868515</td>
      <td>3855.77</td>
      <td>T</td>
    </tr>
    <tr>
      <td>T</td>
      <td>chr9</td>
      <td>PASS</td>
      <td>0/1:99:239,40:279:494,0,6971</td>
      <td>rs34769225</td>
      <td>MLEAC=1;AC=1;BaseQRankSum=1.590;MLEAF=0.500;exons=23;ReadPosRankSum=-1.157;AF=0.500;genes=SETX;AN=2;MQ0=0;FS=0.000;MQ=60.03;DP=331;cDNA_changes=c.6843-5T&gt;-;MQRankSum=1.451;transcripts=NM_015046;dataset=full;protein_changes=NA;ClippingRankSum=0.724;QD=1.38</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>135152543</td>
      <td>456.73</td>
      <td>TA</td>
    </tr>
    <tr>
      <td>TAAA</td>
      <td>chr10</td>
      <td>PASS</td>
      <td>1/1:10:2,26:28:725,10,0</td>
      <td>.</td>
      <td>MLEAC=1;AC=2;BaseQRankSum=0.751;MLEAF=0.500;exons=UTR5;ReadPosRankSum=0.751;AF=1.00;genes=MTPAP;AN=2;MQ0=0;FS=0.000;MQ=60.00;DP=86;cDNA_changes=NA;MQRankSum=-0.751;transcripts=NM_018109;dataset=full;protein_changes=NA;ClippingRankSum=-0.751;QD=0.50</td>
      <td>GT:GQ:AD:DP:PL</td>
      <td>30638208</td>
      <td>129.19</td>
      <td>T</td>
    </tr>
  </tbody>
</table>
    </div>
</body>
</html>